RIP-7 Multiple Directories Storage Support

Jason918 opened the pull request:
**Make sure set the target branch to `develop`**

## What is the purpose of the change

Follow up PR of #3348 
Update according to  comments, see change log

## Brief changelog

1. remove isMultiCommitlogPathEnable
2. In org.apache.rocketmq.store.DefaultMessageStore.CleanCommitLogService#isSpaceToDelete, use minPhysicRatio to determin when the broker's disk is full. And mark all store path when it's full.
3. In org.apache.rocketmq.store.MultiPathMappedFileQueue#tryCreateMappedFile, we ignore readonly paths and  paths marked full in step 2.

## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [x] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

caigy commented on broker/src/main/java/org/apache/rocketmq/broker/processor/SendMessageProcessor.java:
```diff
@@ -635,8 +636,16 @@ public SocketAddress getStoreHost() {
+        if (storePath.contains(MessageStoreConfig.MULTI_PATH_SPLITTER)) {
+            String[] paths = storePath.trim().split(MessageStoreConfig.MULTI_PATH_SPLITTER);
+            for (String storePathPhysic : paths) {
+                physicRatio = Math.min(physicRatio, UtilAll.getDiskPartitionSpaceUsedPercent(storePathPhysic));
```
Is `Math.min()` correct here? `physicRatio` is initialized as `-1` and `UtilAll.getDiskPartitionSpaceUsedPercent()` always returns value greater or equal to `-1`.

Jason918 replied:
@caigy Yes, you are right. Fixed.

caigy commented on store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:
```diff
@@ -782,10 +783,20 @@ private String getStorePathPhysic() {
-            double physicRatio = UtilAll.getDiskPartitionSpaceUsedPercent(getStorePathPhysic());
+        String commitLogStorePath = DefaultMessageStore.this.getMessageStoreConfig().getStorePathCommitLog();
+        if (commitLogStorePath.contains(MessageStoreConfig.MULTI_PATH_SPLITTER)) {
+            double maxValue = Double.MIN_VALUE;
```
Is `maxPhysicsUsedRatio` more appropriate?

Jason918 replied:
@caigy Yes, `maxPhysicsUsedRatio` is better .
But after some consideration, I think we should use `minPhysicsUsedRatio`. It's more useful for admins, because we use min value of all store path to determine whether disk is full and we should clean disk immediately.

caigy replied:
@Jason918 I agree. Using `minPhysicsUsedRatio` is more suitable for this scenario.

francisoliverlee reviewed (changes requested):
checked and muti commit log dir work well. but some errors in DiskCheckScheduledThread and AdminBrokerThread measuring Disk usage.it seems AdminBrokerThread error when no any topics

![image](https://user-images.githubusercontent.com/5908412/133665616-8145be23-db72-4128-b07c-578246611dc4.png)

```
2021-09-17 02:24:50 ERROR DiskCheckScheduledThread1 - Error when measuring disk space usage, file doesn't exist on this path: /opt/rocketmq/master-data1/commitlog:/opt/rocketmq/master-data2/commitlog
2021-09-17 02:25:00 ERROR AdminBrokerThread_7 - Error when measuring disk space usage, file doesn't exist on this path: /opt/rocketmq/master-data1/consumequeue
2021-09-17 02:25:00 ERROR DiskCheckScheduledThread1 - Error when measuring disk space usage, file doesn't exist on this path: /opt/rocketmq/master-data1/commitlog:/opt/rocketmq/master-data2/commitlog
2021-09-17 02:25:03 ERROR AdminBrokerThread_8 - Error when measuring disk space usage, file doesn't exist on this path: /opt/rocketmq/master-data1/consumequeue
2021-09-17 02:25:10 ERROR DiskCheckScheduledThread1 - Error when measuring disk space usage, file doesn't exist on this path: /opt/rocketmq/master-data1/commitlog:/opt/rocketmq/master-data2/commitlog

```

Jason918 commented:
> DiskCheckScheduledThread1

Error in `DiskCheckScheduledThread1` is due to  storePathCommitLog usage in "DefaultMessageStore.CleanCommitLogService#isSpaceFull".  I fixed it and added test in last commit.
Sorry for missing that. I have checked all the usage of `storePathCommitLog`.

Jason918 commented:
> AdminBrokerThread error when no any topics

The log shows that the error in AdminBrokerThread is due to the directory for storing CQs  is not created. 
It's a little bit strange, this PR should have nothing to do with CQs.

Jason918 commented:
@francisoliverlee Added path empty check to avoid error log. Checked with `bin/mqadmin brokerStatus  -n 127.1:9876 -b 127.1:10911`, no error logs appears in broker.log.
Please take a look.

RongtongJin reviewed (approved):
LGTM~, but only one disk is written at the same time, which does not take advantage of the increased write bandwidth after multi-disk deployment.

duhenglucky commented:
> > DiskCheckScheduledThread1
> 
> @francisoliverlee Thank you for the notice.
> Error in `DiskCheckScheduledThread1` is due to storePathCommitLog usage in "DefaultMessageStore.CleanCommitLogService#isSpaceFull". I fixed it and added test in last commit.
> Sorry for missing that. I have checked all the usage of `storePathCommitLog`.

Maybe you discarded this method _org.apache.rocketmq.common.UtilAll#getDiskPartitionSpaceUsedPercent_, or should we check the existence of a directory at the start of the broker?

Jason918 commented on store/src/main/java/org/apache/rocketmq/store/config/MessageStoreConfig.java:
```diff
@@ -17,10 +17,14 @@
 
 public class MessageStoreConfig {
+
+    public static final String MULTI_PATH_SPLITTER = ":";
```
YES, it would be confusing. Thank you for the reminding.
A common reserved character would be a nice fit for this splitter.
According to[ Comparison of file Systems](https://en.wikipedia.org/wiki/Comparison_of_file_systems), different files system have different reserved characters. So there is no best character for this splitter,  and I am using ',' as default splitter, user can change it using system property `rocketmq.broker.multiPathSplitter`

Jason918 commented:
> It seems this error still exists ，and maybe you discarded this method _org.apache.rocketmq.common.UtilAll#getDiskPartitionSpaceUsedPercent_, or should we check the existence of a directory at the start of the broker?

yes, for now, user have to put existing file paths in `storePathCommitLog`.

I think auto creation is better, as we can update it the online.

Jason918 commented:
> LGTM~, but only one disk is written at the same time, which does not take advantage of the increased write bandwidth after multi-disk deployment.

After this multi-disk deployment, read bandwidth will be fully utilized. 
In current one big commit log writing pattern, writing concurrently requires much more work. Maybe soft raid is more suitable for now, if write bandwidth is the physical bottleneck.

ni-ze reviewed (commented):
seems like  muti-commit log dir make sence.

francisoliverlee commented:
> > AdminBrokerThread error when no any topics
> 
> @francisoliverlee
> The log shows that the error in AdminBrokerThread is due to the directory for storing CQs is not created.
> It's a little bit strange, this PR should have nothing to do with CQs.
> I would try to reproduce it.

checked and works OK 👍

francisoliverlee commented:
> > > DiskCheckScheduledThread1
> > 
> > 
> > @francisoliverlee Thank you for the notice.
> > Error in `DiskCheckScheduledThread1` is due to storePathCommitLog usage in "DefaultMessageStore.CleanCommitLogService#isSpaceFull". I fixed it and added test in last commit.
> > Sorry for missing that. I have checked all the usage of `storePathCommitLog`.
> 
> It seems this error still exists ，and maybe you discarded this method _org.apache.rocketmq.common.UtilAll#getDiskPartitionSpaceUsedPercent_, or should we check the existence of a directory at the start of the broker?

when doing getDiskPartitionSpaceUsedPercent(),  how about a WARNING for NO consume queue dir ?
@Jason918 @duhenglucky

yuz10 commented on broker/src/main/java/org/apache/rocketmq/broker/processor/SendMessageProcessor.java:
```diff
@@ -635,8 +636,16 @@ public SocketAddress getStoreHost() {
+                physicRatio = Math.min(physicRatio, UtilAll.getDiskPartitionSpaceUsedPercent(storePathPhysic));
+            }
+        } else {
+            physicRatio = UtilAll.getDiskPartitionSpaceUsedPercent(storePath);
```
this if branch is not necessary, could use the same code with MULTI_PATH branch

Jason918 replied:
updated.

yuz10 commented on store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:
```diff
@@ -782,16 +783,28 @@ private String getStorePathPhysic() {
 
-        {
-            double physicRatio = UtilAll.getDiskPartitionSpaceUsedPercent(getStorePathPhysic());
+        String commitLogStorePath = DefaultMessageStore.this.getMessageStoreConfig().getStorePathCommitLog();
```
Dose it support dledge mode? the original code getStorePathPhysic() has different value of  commitLogStorePath in dledge mode

Jason918 replied:
Thank you for the reminding. 
Changed to `getStorePathPhysic()`

yuz10 commented on store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:
```diff
@@ -1650,25 +1663,49 @@ private boolean isSpaceToDelete() {
+                if (commitLogStorePath.contains(MessageStoreConfig.MULTI_PATH_SPLITTER)) {
+                    storePaths = commitLogStorePath.trim().split(MessageStoreConfig.MULTI_PATH_SPLITTER);
+                } else {
+                    storePaths = new String[]{commitLogStorePath};
```
not necessary if branch, could use the same code with MULTI_PATH branch

Jason918 replied:
updated.

yuz10 commented on store/src/main/java/org/apache/rocketmq/store/MappedFileQueue.java:
```diff
@@ -144,35 +145,39 @@ void deleteExpiredFile(List<MappedFile> files) {
-                }
+    public boolean doLoad(List<File> files) {
+        // ascending order
+        Collections.sort(files);
```
According to the code :
![image](https://user-images.githubusercontent.com/14816818/134317644-4061cb3f-5470-43bf-bf52-ee974458db10.png)
if paths are a/, b/ then files will be created such as:
a/000000   b/000001  a/000002  b/000003
so sort these can cause wrong order.

Jason918 replied:
Great point.
Changed `Collections.sort(files)` to `files.sort(Comparator.comparing(File::getName))`,  sort the file using simple filenames, aka, the last part of the path.

And added test in org.apache.rocketmq.store.MultiPathMappedFileQueueTest#testLoadReadOnlyMappedFiles， PTAL.

lollipopjin commented on store/src/main/java/org/apache/rocketmq/store/MultiPathMappedFileQueue.java:
```diff
@@ -0,0 +1,127 @@
+        //do not create file is space is nearly full.
+        availableStorePath.removeAll(fullStorePaths);
+
+        //if no store path left, fall back to wriable store path.
```
wriable --> writable

Jason918 replied:
fixed.

lollipopjin commented:
If we add a cloud disk to the running ecs, can the new cloud disk be recognized by updating broker.conf dynamically?

Jason918 commented:
> If we add a cloud disk to the running ecs, can the new cloud disk be recognized by updating broker.conf dynamically?

Yes, It should work. You can add path after `storePathCommitLog`. 
But, one thing to be noticed, if your server is started with only one commit log path, you have to restart the server to change it to multi-path-commit-log mode. Because `mappedFileQueue` in class `Commitlog` is initialized during start up.
One way to avoid this restart is to add a `MessageStoreConfig#MULTI_PATH_SPLITTER`(default is ',' and can be changed by system property "rocketmq.broker.multiPathSplitter") after origin single storePathCommitLog, e.g. "~/store/commitlog,".

francisoliverlee reviewed (approved):
OK

cserwen commented:
`readOnlyCommitLogStorePaths`: I have a question to ask. What is the meaning of this configuration, is it used for smooth upgrade? @Jason918

Jason918 commented:
> `readOnlyCommitLogStorePaths`: I have a question to ask. What is the meaning of this configuration, is it used for smooth upgrade? @Jason918

Yep, it can be used to decommission some broken disks.
